### PR TITLE
arrow timestamp to ast timestamp conversion support

### PR DESCRIPTION
**Pull request by sgrebnov**
Addresses the following issues:
https://github.com/spiceai/spiceai/issues/1676
https://github.com/spiceai/spiceai/issues/1625


Dremio example
```
# select * from taxi_trips where pickup_datetime = '2013-08-22 06:24:12-02'
# +---------------------+-----------------+------------------+-------------+------------+--------------+
# | pickup_datetime     | passenger_count | trip_distance_mi | fare_amount | tip_amount | total_amount |
# +---------------------+-----------------+------------------+-------------+------------+--------------+
# | 2013-08-22T08:24:12 | 1               | 1.1              | 7.5         | 0.0        | 8.0          |
# +---------------------+-----------------+------------------+-------------+------------+--------------+
```

```
sql> explain select * from taxi_trips where pickup_datetime = '2013-08-22 06:24:12-02'
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Federated                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|               |  Projection: datasets.taxi_trips.pickup_datetime, datasets.taxi_trips.passenger_count, datasets.taxi_trips.trip_distance_mi, datasets.taxi_trips.fare_amount, datasets.taxi_trips.tip_amount, datasets.taxi_trips.total_amount                                                                                                                                                                                                                                      |
|               |   Filter: CAST(datasets.taxi_trips.pickup_datetime AS Timestamp(Nanosecond, None)) = CAST(Utf8("2013-08-22 06:24:12-02") AS Timestamp(Nanosecond, None))                                                                                                                                                                                                                                                                                                            |
|               |     TableScan: datasets.taxi_trips                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| physical_plan | SchemaCastScanExec                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |   RepartitionExec: partitioning=RoundRobinBatch(14), input_partitions=1                                                                                                                                                                                                                                                                                                                                                                                             |
|               |     VirtualExecutionPlan name=dremio compute_context=url=grpc://20.163.171.8:32010,username=demo sql=SELECT datasets.taxi_trips.pickup_datetime, datasets.taxi_trips.passenger_count, datasets.taxi_trips.trip_distance_mi, datasets.taxi_trips.fare_amount, datasets.taxi_trips.tip_amount, datasets.taxi_trips.total_amount FROM datasets.taxi_trips WHERE (CAST(datasets.taxi_trips.pickup_datetime AS TIMESTAMP) = CAST('2013-08-22 06:24:12-02' AS TIMESTAMP)) |
|               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
``` 

Dremio does not support  TIMESTAMPTZ so the following won't work as now() type is Timestamp(Millisecond, Some("+00:00")).

```
sql> explain select * from taxi_trips where pickup_datetime < now()
+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Federated                                                                                                                                                                                                                                                                                                                                                                                                                                            |
|               |  Projection: datasets.taxi_trips.pickup_datetime, datasets.taxi_trips.passenger_count, datasets.taxi_trips.trip_distance_mi, datasets.taxi_trips.fare_amount, datasets.taxi_trips.tip_amount, datasets.taxi_trips.total_amount                                                                                                                                                                                                                       |
|               |   Filter: CAST(datasets.taxi_trips.pickup_datetime AS Timestamp(Millisecond, Some("+00:00"))) < CAST(now() AS Timestamp(Millisecond, Some("+00:00")))                                                                                                                                                                                                                                                                                                |
|               |     TableScan: datasets.taxi_trips                                                                                                                                                                                                                                                                                                                                                                                                                   |
| physical_plan | SchemaCastScanExec                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|               |   RepartitionExec: partitioning=RoundRobinBatch(14), input_partitions=1                                                                                                                                                                                                                                                                                                                                                                              |
|               |     VirtualExecutionPlan name=dremio compute_context=url=grpc://20.163.171.8:32010,username=demo sql=SELECT datasets.taxi_trips.pickup_datetime, datasets.taxi_trips.passenger_count, datasets.taxi_trips.trip_distance_mi, datasets.taxi_trips.fare_amount, datasets.taxi_trips.tip_amount, datasets.taxi_trips.total_amount FROM datasets.taxi_trips WHERE (CAST(datasets.taxi_trips.pickup_datetime AS TIMESTAMPTZ) < CAST(now() AS TIMESTAMPTZ)) |
|               |                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### CODE REVIEW

File: datafusion/sql/src/unparser/expr.rs [hunk 1]
```diff
@@ -958,8 +959,16 @@ impl Unparser<'_> {
             }
             DataType::Float32 => Ok(ast::DataType::Float(None)),
             DataType::Float64 => Ok(ast::DataType::Double),
-            DataType::Timestamp(_, _) => {
-                not_impl_err!("Unsupported DataType: conversion: {data_type:?}")
+            DataType::Timestamp(time_unit, _) => {
+
+                let time_precision = match time_unit {
```
Comment (sgrebnov) [inline datafusion/sql/src/unparser/expr.rs:964]: Not 100% sure time_precision should be specified. Seems adding precision is the right thing as we have this info.
Not all precision is supported by target engines, for example  despite the following sql works in Postgres `select CAST('2013-08-22 06:24:12-02' AS TIMESTAMP(9))` I see `TIMESTAMP(9) precision reduced to maximum allowed, 6` in Postgres logs. Also tested max precision in Dremio, Snowflake - works fine.

**Comment by sgrebnov** [comment]
Converted to draft as snowflake requires TIMESTAMPTZ to work correct with the time zone. Dremio requires TIMESTAMPTZ. Will be looking how to fix this in Spice where we know the engine

File: datafusion/sql/src/unparser/expr.rs [hunk 2]
```diff
@@ -958,8 +958,14 @@ impl Unparser<'_> {
             }
             DataType::Float32 => Ok(ast::DataType::Float(None)),
             DataType::Float64 => Ok(ast::DataType::Double),
-            DataType::Timestamp(_, _) => {
-                not_impl_err!("Unsupported DataType: conversion: {data_type:?}")
+            DataType::Timestamp(_, tz) => {
+
+                let tz_info = match tz {
+                    Some(_) => TimezoneInfo::Tz,
```
Comment (phillipleblanc) [inline datafusion/sql/src/unparser/expr.rs:964]: Tz is specific to the Postgres dialect: https://github.com/sqlparser-rs/sqlparser-rs/blob/main/src/ast/data_type.rs#L632

To make this compatible with more systems, we should use `WithTimeZone`

Comment (sgrebnov) [reply datafusion/sql/src/unparser/expr.rs]: Updated, thank you

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**